pkg/operator/terminationobserver/termination_observer.go

```diff
@@ -0,0 +1,249 @@
+package terminationobserver
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+var (
+	controllerWorkQueueKey = "key"
+
+	// terminationEventReasons lists all events that are observed when an API server is shutting down gracefully.
+	terminationEventReasons = []string{
+		"TerminationStart",
+		"TerminationPreShutdownHooksFinished",
+		"TerminationPreShutdownHooksFinished",
+		"TerminationMinimalShutdownDurationFinished",
+		"TerminationStoppedServing",
+		"TerminationGracefulTerminationFinished",
+	}
+)
+
+type TerminationObserver struct {
+	targetNamespace string
+
+	podsGetter corev1client.PodsGetter
+
+	cachesToSync  []cache.InformerSynced
+	queue         workqueue.RateLimitingInterface
+	eventRecorder events.Recorder
+
+	apiServers      []*apiServerObserver
+	apiServersMutex sync.RWMutex
+}
+
+type apiServerObserver struct {
+	serverName            string
+	lastCreationTimestamp time.Time
+	terminationEvents     []*corev1.Event
+}
+
+func NewTerminationObserver(
+	targetNamespace string,
+	kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	podsGetter corev1client.PodsGetter,
+	eventRecorder events.Recorder,
+) *TerminationObserver {
+	c := &TerminationObserver{
+		targetNamespace: targetNamespace,
+		podsGetter:      podsGetter,
+		eventRecorder:   eventRecorder.WithComponentSuffix("termination-observer"),
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "TerminationObserver"),
+	}
+
+	kubeInformersForTargetNamespace.Core().V1().Pods().Informer().AddEventHandler(c.eventHandler())
+	kubeInformersForTargetNamespace.Core().V1().Events().Informer().AddEventHandler(c.terminationEventRecorder())
+
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Pods().Informer().HasSynced)
+	c.cachesToSync = append(c.cachesToSync, kubeInformersForTargetNamespace.Core().V1().Events().Informer().HasSynced)
+
+	return c
+}
+
+// updateApiServerCreationTimestamp updates the creation timestamp for single API server
+func (c *TerminationObserver) updateApiServerCreationTimestamp(name string, lastCreationTimestamp time.Time) {
+	c.apiServersMutex.Lock()
```

Comment:
this makes me nervous.  Surely this isn't so high throughput that it needs to run multi-threaded

Reply:
@deads2k i'm sure the unit tests when ran with `-race` will implode without this locking.

Reply:
+1, this looks suspicious. Forbid more than one worker.